[tools] Add copyright lint tooling for Golang files

sgnn7 opened the pull request:
### What does this PR do?

This tool addition can be used via `inv lint-copyrights` which traverses the repo
tree and fails if it finds any Golang files with an unexpected copyright
header.

For now the tool is intended to be invoked manually but once all the files
get fixed, we can add it to our internal CI/CD pipeline as an automated
check.

### Motivation

Legal compliance needs

### Additional Notes

- Eventually we would like to enable this check in CI/CD but we need all the
  failures cleaned up first.
  
### Possible Drawbacks / Trade-offs

The tool could ideally have the ability to auto-fix problems but that's outside
of the scope of this task right now.

### Describe how to test/QA your changes

N/A but if test is needed, running `inv lint-copyrights` should show the output.

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [x] The appropriate `team/..` label has been applied, if known.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

sgnn7 committed:
[tools] Add copyright linting tool for Golang files

This addition can be used via `inv lint-copyright` which traverses the repo
tree and fails if it finds any Golang files with an unexpected copyright
header.

For now the tool is intended to be invoked manually but once all the files
get fixed, we can add it to our internal CI/CD pipeline as an automated
check.

mx-psi commented on tasks/libs/copyright.py:
```diff
@@ -0,0 +1,184 @@
+            failing_files.append(filepath)
+
+        total_files = len(files)
+        if failing_files != 0:
```
failing_files is a list, maybe you wanted to do this?
```suggestion
        if failing_files:
```

sgnn7 replied:
Yup - good catch!

mx-psi commented on tasks/test.py:
```diff
@@ -548,6 +549,16 @@ def lint_python(ctx):
+@task
+def lint_copyrights(_):
+    """
+    Checks that all files matching the search pattern (default: '**/*.go')
+    contain the appropriate copyright header.
```
Guessing this used to have a 'search pattern' parameter and it doesn't anymore?
```suggestion
    Checks that all Go files contain the appropriate copyright header.
```

sgnn7 replied:
It's the other way around. For now this only checks Golang files but we may expand it to do the others types like Python and C. I'll apply your suggestion for now though.

mx-psi replied:
ah, right, thanks for explaining!

mx-psi commented on tasks/libs/copyright.py:
```diff
@@ -0,0 +1,184 @@
+        # `all_matching_files` is a generator so we have to do the counting ourselves
+        all_matching_files_cnt = 0
+
+        filtered_files = []
```
nit: To me it would feel more 'Pythonic' to define an `is_excluded_file` predicate and then do
```python
filtered_files = [filepath for filepath in all_matching_files if is_excluded_file(filepath)]
```

sgnn7 replied:
@mx-psi I made the inside loop be predicate-based but due to the generator being the iterable over which we evaluate this block and the fact that we need the total count for debugging, I left the outer loop as-is. I know the generator can be turned into a list easy for the length calculation but I think at that point we would be skyrocketing in mem usage so I opted against it.

mx-psi replied:
Makes sense, I think the predicate improvement makes this look better just by itself :)

sgnn7 committed:
[tools] Fix minor nits from code review

There were some inefficiencies and nits in the original commit that
needed fixes so this change fixes them.

sgnn7 committed:
[tools] Add loglevel prefix to most copyright linter printouts

This should make it visually easier to parse the output of this tooling.
